spec/support/analytics_recording_helper.rb

```diff
@@ -0,0 +1,466 @@
+        # tokenizer. But if {tokenizer} returns a Symbol / String, we use that
+        # as the namespace.
+        # This is because some IDs are safe to put in a global namespace
+        # because we won't have collisions (like UUIDs). Other things
+        # (like database ids) aren't unique enough, so we need a hint
+        # as to their context (essentially "what database table is this
+        # numeric ID from?")
+
+        token_type = tokenizer_result == true ?
+          tokenizer_name.to_s :
+          tokenizer_result.to_s
+
+        state[:tokenizers] ||= {}
+        state[:tokenizers][token_type] ||= {}
+
+        tokens = state[:tokenizers][token_type]
+
+        if !tokens.has_key?(value)
+          tokens[value] = "#{token_type}:#{tokens.count + 1}"
+        end
+
+        return tokens[value]
+      end
+
+      value
+    },
+  }.freeze
+
+  # These are used when generating a friendly summary of an event to provide
+  # short bits of additional context.
+  EVENT_CONTEXT_GIVERS = [
+    ->(event) {
+      event.dig(:properties, :event_properties, :action)
+    },
+    ->(event) {
+      success = event.dig(:properties, :event_properties, :success)
+      if success == true
+        'succeeded'
+      elsif success == false
+        'failed'
+      end
+    },
+  ].freeze
+
+  # Either:
+  # 1) Records analytics events logged during {block} when the RECORD_ANALYTICS env is truthy OR
+  # 2) Verifies analytics events logged during {block} against previously-recorded events
+  # @param {String} file_name File to record analytics events to
+  # @return {Symbol,nil} Either :recorded (if analytics were recorded) or :checked (if events were
+  #                      checked against a previous recording. If no recorded events could b
+  #                      found to check against, returns nil.
+  def record_or_verify_analytics(
+    file_name:,
+    &block
+  )
+    if should_record_analytics?
+      record_analytics(file_name:, &block)
+      return :recorded
+    end
+
+    actual_events = []
+    middleware = proc { |event|
+      event.tap { actual_events << event }
+    }
+
+    Analytics.with_default_middleware(middleware) do
+      block.call
+    end
+
+    expected_events = begin
+      load_analytics_events_from_file(file_name:)
+    rescue Errno::ENOENT
+      return nil
+    end
+
+    assert_logged_analytics_events_match(
+      actual_events:,
+      expected_events:,
+      file_name:,
+    )
+
+    :checked
+  end
+
+  # Verifies that {actual_events} matches {expected_events}.
+  # Order matters, but some wiggle room is allowed via the {window_size} parameter.
+  # @param [Hash[]] actual_events Array of actual events logged.
+  # @param [Hash[]] expected_events Array of analytics event hashes.
+  # @param [String,nil] file_name File name events were loaded from (for error reporting)
+  # @param [Number] window_size Number of events we look at when trying to find a match.
+  def assert_logged_analytics_events_match(
```

Comment:
I think this is more thorough than our current analytics feature spec, since it is requiring that every event seen match 1:1 with an event that was previously recorded.

The current spec defines a set of events, and then asserts that each one was logged. If the app starts firing a new analytics event, that won't cause the analytics spec to fail--it needs to be manually updated to know about the new event.